src/utils/ICS24Host.sol

```diff
@@ -137,21 +112,28 @@ library ICS24Host {
     function packetCommitmentBytes32(IICS26RouterMsgs.Packet memory packet) internal pure returns (bytes32) {
         return sha256(
             abi.encodePacked(
-                SafeCast.toUint64(uint256(packet.timeoutTimestamp) * 1_000_000_000),
-                uint64(0),
-                uint64(0),
-                sha256(packet.data),
-                packet.destPort,
-                packet.destChannel
+                packet.timeoutTimestamp, sha256(abi.encodePacked(packet.destChannel)), hashPayload(packet.payloads[0])
             )
         );
     }
 
+    function hashPayload(IICS26RouterMsgs.Payload memory data) internal pure returns (bytes32) {
+        bytes memory buf = abi.encodePacked(
+            sha256(abi.encodePacked(data.sourcePort)),
+            sha256(abi.encodePacked(data.destPort)),
+            sha256(data.value),
+            sha256(abi.encodePacked(data.encoding)),
+            sha256(abi.encodePacked(data.version))
+        );
+
+        return sha256(buf);
+    }
+
     /// @notice Get the packet acknowledgement commitment bytes.
     /// @param ack The acknowledgement to get the commitment for
     /// @return The commitment bytes
     function packetAcknowledgementCommitmentBytes32(bytes memory ack) internal pure returns (bytes32) {
-        return sha256(ack);
+        return sha256(abi.encodePacked(sha256(ack)));
```

Comment:
This is happening because of the multi packet payload:
```go
// CommitAcknowledgement returns the hash of the acknowledgement data.
func CommitAcknowledgement(acknowledgement Acknowledgement) []byte {
	var buf []byte
	for _, ack := range acknowledgement.GetAcknowledgementResults() {
		hash := sha256.Sum256(ack.RecvPacketResult.GetAcknowledgement())
		buf = append(buf, hash[:]...)
	}

	hash := sha256.Sum256(buf)
	return hash[:]
}
```

Reply:
abi.encodePacked is used because sha256 returns bytes32, but only accepts bytes 🤷

Reply:
I think we should define the `Acknowledgement` struct in that case and add an assertion that `.length == 1`. Also, might be possible to wrap in `bytes()` instead of `abi.encodePacked()`, similar to a string?